### PR TITLE
Feature/onclick

### PR DESCRIPTION
**Pull request by Andries-Smit**

### CODE REVIEW

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
+import { Component, DOM } from "react";
+
+import * as classNames from "classnames";
+import { Circle } from "progressbar.js";
+
+import "../ui/ProgressCircle.css";
+
+export interface OnclickProps {
+    onClickType: ProgressOnclick;
+    microflowProps?: MicroflowProps;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: inline MicroflowProps, its used nowhere else

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    pageProps?: PageProps;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Inline PageProps  its used nowhere else

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.xml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            <description>Animates when the progress circle is loaded and updated</description>
+        </property>
+        <property key="progressAttribute" type="attribute" required="true">
+            <caption>Progress attribute</caption>
+            <category>Data</category>
+            <description>Holds the progress value</description>
+            <attributeTypes>
+                <attributeType name="Integer" />
+                <attributeType name="Decimal" />
+            </attributeTypes>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/ProgressCircle.xml]: Missing type Long

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        </property>
+        <property key="maximumValueAttribute" type="attribute" required="false">
+            <caption>Maximum value attribute</caption>
+            <category>Data</category>
+            <description>Highest progress value. Larger then 0 If not set, maximum value defaults to 100</description>
+            <attributeTypes>
+                <attributeType name="Integer" />
+                <attributeType name="Decimal" />
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/ProgressCircle.xml]: Missing type Long

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        "declaration": false,
+        "removeComments": true,
+        "noLib": false,
+        "watch": false,
+        "forceConsistentCasingInFileNames": true,
+        "noFallthroughCasesInSwitch": true,
+        "noImplicitAny": true,
+        "noImplicitReturns": true,
+        "noImplicitThis": true,
+        "skipLibCheck": true
```
Comment (Andries-Smit) [inline tsconfig.json]: Add strictNullChecks

File: webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+const ExtractTextPlugin = require("extract-text-webpack-plugin");
+
+module.exports = {
+    entry: "./src/com/mendix/widget/ProgressCircle/ProgressCircle.ts",
+    output: {
+        path: path.resolve(__dirname, "dist/tmp"),
+        filename: "src/com/mendix/widget/ProgressCircle/ProgressCircle.js",
+        libraryTarget:  "umd",
+        umdNamedDefine: true,
+        library: "com.mendix.widget.ProgressCircle.ProgressCircle"
```
Comment (Andries-Smit) [inline webpack.config.js]: Remove umdNamedDefine and library

File: package.json [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    "webpack": "^1.13.2",
+    "webpack-dev-server": "^1.16.2"
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "progressbar.js": "^1.0.1",
+    "react": "~15.3.2",
+    "react-dom": "~15.3.2"
+  }
+}
```
Comment (Andries-Smit) [inline package.json]: Please update all dependencies

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    },
+    errorDetails: true,
+    module: {
+        loaders: [
+            { test: /\.ts$/, loader: "ts-loader" },
+            { test: /\.json$/, loader: "json" },
+            { test: /\.css$/, loader: ExtractTextPlugin.extract("style-loader", "css-loader") }
+        ],
+        postLoaders: [ {
```
Comment (Andries-Smit) [inline webpack.config.js]: Move the postLoaders to the Karma config with the CLI parameter `codeCoverage`

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                        `Error while executing microflow: ${props.microflowProps.microflow}: ${error.message}`
+                    ),
+                params: {
+                    applyto: "selection",
+                    guids: [ props.microflowProps.guid ]
+                }
+            });
+        } else if (props.onClickType === "showPage" && props.pageProps.page && props.pageProps.guid) {
+            let context = new window.mendix.lib.MxContext();
+            context.setTrackId(props.pageProps.guid);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Mendix client typing does not support `window.mendix.lib.MxContext()` 
Please fix and create a pull request

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 1 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Removed. Wonder who came up with this MicroflowProps thing... removed it before I saw this.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 2 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: I don't have OnclickProps anymore either. Inlined it all.

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.xml (hunk 3 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/ProgressCircle.xml]: Added.

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.xml (hunk 4 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/ProgressCircle.xml]: Added.

File: tsconfig.json (hunk 5 above)
Comment (edwinmp) [reply tsconfig.json]: Added.

File: webpack.config.js (hunk 6 above)
Comment (edwinmp) [reply webpack.config.js]: removed.

File: webpack.config.js (hunk 8 above)
Comment (edwinmp) [reply webpack.config.js]: Done.

File: package.json (hunk 7 above)
Comment (edwinmp) [reply package.json]: Done, except for react.

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+        return true;
+    }
+
+    private updateRendering() {
+        render(createElement(Circle, {
+            animate: this.animate,
+            contextObject: this.contextObject,
+            maximumValue: this.contextObject && this.maximumValueAttribute
+                ? Number(this.contextObject.get(this.maximumValueAttribute))
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/ProgressCircle.ts]: Explicit conversion is not need, just use `? his.contextObject.get(this.maximumValueAttribute) as number`

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                : undefined,
+            microflow: this.microflow,
+            onClickType: this.onClickEvent,
+            page: this.page,
+            pageSettings: this.pageSettings,
+            textSize: this.textSize,
+            value: this.contextObject ? Number(this.contextObject.get(this.progressAttribute)) : null
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/ProgressCircle.ts]: Explicit conversion is not need, just use `? this.contextObject.get(this.maximumValueAttribute) as number`

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+    private createProgressCircle() {
+        this.progressCircle = new Circle(this.progressNode, {
+            duration: this.props.animate ? 800 : -1,
+            strokeWidth: 6,
+            trailWidth: 6
+        });
+        this.progressCircle.trail.className.baseVal = "widget-progress-circle-trail-path";
+        this.progressCircle.path.className.baseVal = "widget-progress-circle-path";
+    }
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Suggestion; Create pull request for progressbar.js where the class names could be set via the API.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                    applyto: "selection",
+                    guids: [ contextObject.getGuid() ]
+                }
+            });
+        } else if (contextObject && onClickType === "showPage" && page && contextObject.getGuid()) {
+            let context = new window.mendix.lib.MxContext();
+            context.setTrackId(contextObject.getGuid());
+            context.setTrackEntity(contextObject.getEntity());
+
+            window.mx.ui.openForm(page, {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: The `openPage` also an `error` callback. Which less likely to throw an error, but for good practice, we should add it.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        let progress = 0;
+        let progressText: string;
+        if (value === null || typeof value === "undefined") {
+            progressText = "--";
+        } else if (maximum <= 0) {
+            window.console.warn("The maximum value is 0. Progress is set to Invalid");
+            progressText = "Invalid";
+        } else {
+            progress = Math.round((value / maximum) * 100);
+            progressText = progress + "%";
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Suggestion API; for a more uniform way of creating the progress text would be a Dynamic text (like we have on captions of the buttons would be ideal.) Maybe it would be for plugin widgets.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+                className: classNames(`widget-progress-circle-${this.props.textSize}`,
+                    {
+                        negative: this.props.value < 0,
+                        "red-progress-text": this.props.maximumValue < 1
+                    }
+                ),
+                onClick: () => this.handleOnClick(),
+                ref: node => this.progressNode = node
+            }),
+            this.state.alertMessage ? createElement(Alert, { message: this.state.alertMessage }) : null
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Can we move the conditional statement to the `Alert` component?

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+describe("ProgressCircle", () => {
+
+    beforeAll(() => {
+        window.mx = mockMendix;
+        window.mendix = { lib: { MxContext: MockContext } };
+    });
+
+    let progressCircle: progressbar.Circle;
+    const render = (props: ProgressCircleProps, click: boolean = false) => {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Name `render` does not represent its content, should it be `renderDoClick`? or 2 different functions.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            strokeWidth: 6,
+            trailWidth: 6
+        });
+        this.progressCircle.trail.className.baseVal = "widget-progress-circle-trail-path";
+        this.progressCircle.path.className.baseVal = "widget-progress-circle-path";
+    }
+
+    private setProgress(value: number | null, maximum: number = 100) {
+        if (!this.progressCircle) {
+            this.createProgressCircle();
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Probably not the right place to call the create. Move to `componentDidMount`

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+        expect(progressCircle.text.textContent).toBe("-1%");
+    });
+
+    it("renders a circle with text greater than 100% when the value is greater than the maximum", () => {
+        spyOnCircle();
+
+        const progress = newCircleInstance({ value: 180 });
+        progress.componentDidMount();
+
+        expect(progressCircle.text.textContent).toBe("180%");
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Expect the text to be `180%` but the circle progress value only the `1`. Please check both.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        } else if (maximum <= 0) {
+            window.console.warn("The maximum value is 0. Progress is set to Invalid");
+            progressText = "Invalid";
+        } else {
+            progress = Math.round((value / maximum) * 100);
+            progressText = progress + "%";
+        }
+
+        let animateValue = progress / 100;
+        animateValue = animateValue <= 1
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Suggestion: create pull fix for the progressbar.js This behaviour is generic.

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+        spyOnCircle();
+
+        const progress = newCircleInstance({ value: 80 });
+        progress.componentDidMount();
+        progress.componentWillUnmount();
+
+        expect(destroy).toHaveBeenCalled();
+    });
+
+    it("renders a circle no text when no value is specified", () => {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: no text or "--" ?

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+    });
+
+    describe("configured to call a microflow on click", () => {
+        const contextObject: any = {
+            getGuid: () => { return random.uuid() },
+            getEntity: () => { return random.uuid() }
+        };
+        const circleProps: ProgressCircleProps = {
+            contextObject,
+            microflow: "IVK_Onclick",
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: New Mendix standard prefix action with `ACT_`

File: tests/matchers.ts [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+                    pass: true
+                };
+            }
+        }
+    };
+}
+
+beforeEach(() =>
+    jasmine.addMatchers({ toBeElement, toMatchStructure, toHaveClass })
+);
```
Comment (Andries-Smit) [inline tests/matchers.ts]: Missing new line and file

File: tslint.json [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        ],
+        "curly": false
+    },
+    "jsRules": {
+        "trailing-comma": [ false ],
+        "object-literal-sort-keys": false,
+        "array-bracket-spacing": [ true, "always" ]
+    }
+}
```
Comment (Andries-Smit) [inline tslint.json]: missing newline end file

File: typings/core-js.d.ts [hunk 24]
```diff
@@ -0,0 +1,71 @@
+// Type definitions for core-js v0.9.7
+declare type PropertyKey = string | number | symbol;
```
Comment (Andries-Smit) [inline typings/core-js.d.ts]: Is this file still required, with the latest version of 'core-js 'and '@types/core-js' is stalled?

**Review by Andries-Smit — COMMENTED** [review]
Please follow up on the comments

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.ts (hunk 10 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/ProgressCircle.ts]: Done.

File: src/com/mendix/widget/ProgressCircle/ProgressCircle.ts (hunk 11 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/ProgressCircle.ts]: Done.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 15 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Not without turning it into a class or at least disrupting it's current tranquility as a pure stateless component.
Alternative is to always render it and control its visibility based on the `hidden` class. Don't think it's faithful to react to do it this way though.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 17 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Moved.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 13 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Added.

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts (hunk 16 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Separated into two functions.

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts (hunk 20 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Fixed.

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts (hunk 18 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Couldn't test this. Regardless of the value, `progressCircle.value()` returned `NaN`.
What I did was:
```
spyOn(progressCircle, "animate").and.callThrough();
expect(progressCircle.animate).toHaveBeenCalledWith(1);
```

File: src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts (hunk 21 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/__tests__/ProgressCircle.spec.ts]: Changed.

File: tests/matchers.ts (hunk 22 above)
Comment (edwinmp) [reply tests/matchers.ts]: Copy/paste mishaps :D

File: tslint.json (hunk 23 above)
Comment (edwinmp) [reply tslint.json]: Fixed.

File: typings/core-js.d.ts (hunk 24 above)
Comment (edwinmp) [reply typings/core-js.d.ts]: Errors still occur after update.

File: karma.conf.js [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+        "react/lib/ExecutionEnvironment",
+        "react/lib/ReactContext",
+        "react/addons",
+        "jsdom"
+    ])
+});
+
+module.exports = function(config) {
+    if (config.codeCoverage) {
+        console.log("With instrumenter");
```
Comment (Andries-Smit) [inline karma.conf.js]: Remove `console.log`

File: Gruntfile.js [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+                "./dist/MxTestProject/widgets/" + pkg.name + ".mpk"
+            ]
+        }
+    });
+
+    grunt.loadNpmTasks("grunt-contrib-compress");
+    grunt.loadNpmTasks("grunt-contrib-clean");
+    grunt.loadNpmTasks("grunt-contrib-watch");
+    grunt.loadNpmTasks("grunt-contrib-copy");
+    grunt.loadNpmTasks("grunt-webpack");
```
Comment (Andries-Smit) [inline Gruntfile.js]: Add dependency check

File: karma.conf.js [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+        concurrency: Infinity,
+        coverageReporter: {
+            dir: "./dist/testresults",
+            reporters: [
+                { type: "json", subdir: ".", file: "coverage.json" },
+                { type: "text" }
+            ]
+        }
+    });
+};
```
Comment (Andries-Smit) [inline karma.conf.js]: Missing newline at end of file

File: src/com/mendix/widget/ProgressCircle/components/Alert.ts [hunk 28]
```diff
@@ -0,0 +1,6 @@
+import { DOM, StatelessComponent } from "react";
+
+export const Alert: StatelessComponent<{ message?: string }> = (props) =>
+    props.message
+        ? DOM.div({ className: "alert alert-danger widget-progress-circle-alert" }, props.message)
+        : null as any;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressCircle/components/Alert.ts]: Remove `as any` update your react typings

**Review by Andries-Smit — COMMENTED** [review]
Last remarks.

File: karma.conf.js (hunk 25 above)
Comment (edwinmp) [reply karma.conf.js]: Done.

File: karma.conf.js (hunk 27 above)
Comment (edwinmp) [reply karma.conf.js]: Added.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 9 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Pull request from Guy is pending. Changed the `package.json` link to point to that till the pull request is accepted.

File: src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts (hunk 15 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressCircle/components/ProgressCircle.ts]: Done.
